<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>Format data</title>

<script type="text/javascript">
window.onload = function() {
  var imgs = document.getElementsByTagName('img'), i, img;
  for (i = 0; i < imgs.length; i++) {
    img = imgs[i];
    // center an image if it is the only element of its parent
    if (img.parentElement.childElementCount === 1)
      img.parentElement.style.textAlign = 'center';
  }
};
</script>

<!-- Styles for R syntax highlighter -->
<style type="text/css">
   pre .operator,
   pre .paren {
     color: rgb(104, 118, 135)
   }

   pre .literal {
     color: #990073
   }

   pre .number {
     color: #099;
   }

   pre .comment {
     color: #998;
     font-style: italic
   }

   pre .keyword {
     color: #900;
     font-weight: bold
   }

   pre .identifier {
     color: rgb(0, 0, 0);
   }

   pre .string {
     color: #d14;
   }
</style>

<!-- R syntax highlighter -->
<script type="text/javascript">
var hljs=new function(){function m(p){return p.replace(/&/gm,"&amp;").replace(/</gm,"&lt;")}function f(r,q,p){return RegExp(q,"m"+(r.cI?"i":"")+(p?"g":""))}function b(r){for(var p=0;p<r.childNodes.length;p++){var q=r.childNodes[p];if(q.nodeName=="CODE"){return q}if(!(q.nodeType==3&&q.nodeValue.match(/\s+/))){break}}}function h(t,s){var p="";for(var r=0;r<t.childNodes.length;r++){if(t.childNodes[r].nodeType==3){var q=t.childNodes[r].nodeValue;if(s){q=q.replace(/\n/g,"")}p+=q}else{if(t.childNodes[r].nodeName=="BR"){p+="\n"}else{p+=h(t.childNodes[r])}}}if(/MSIE [678]/.test(navigator.userAgent)){p=p.replace(/\r/g,"\n")}return p}function a(s){var r=s.className.split(/\s+/);r=r.concat(s.parentNode.className.split(/\s+/));for(var q=0;q<r.length;q++){var p=r[q].replace(/^language-/,"");if(e[p]){return p}}}function c(q){var p=[];(function(s,t){for(var r=0;r<s.childNodes.length;r++){if(s.childNodes[r].nodeType==3){t+=s.childNodes[r].nodeValue.length}else{if(s.childNodes[r].nodeName=="BR"){t+=1}else{if(s.childNodes[r].nodeType==1){p.push({event:"start",offset:t,node:s.childNodes[r]});t=arguments.callee(s.childNodes[r],t);p.push({event:"stop",offset:t,node:s.childNodes[r]})}}}}return t})(q,0);return p}function k(y,w,x){var q=0;var z="";var s=[];function u(){if(y.length&&w.length){if(y[0].offset!=w[0].offset){return(y[0].offset<w[0].offset)?y:w}else{return w[0].event=="start"?y:w}}else{return y.length?y:w}}function t(D){var A="<"+D.nodeName.toLowerCase();for(var B=0;B<D.attributes.length;B++){var C=D.attributes[B];A+=" "+C.nodeName.toLowerCase();if(C.value!==undefined&&C.value!==false&&C.value!==null){A+='="'+m(C.value)+'"'}}return A+">"}while(y.length||w.length){var v=u().splice(0,1)[0];z+=m(x.substr(q,v.offset-q));q=v.offset;if(v.event=="start"){z+=t(v.node);s.push(v.node)}else{if(v.event=="stop"){var p,r=s.length;do{r--;p=s[r];z+=("</"+p.nodeName.toLowerCase()+">")}while(p!=v.node);s.splice(r,1);while(r<s.length){z+=t(s[r]);r++}}}}return z+m(x.substr(q))}function j(){function q(x,y,v){if(x.compiled){return}var u;var s=[];if(x.k){x.lR=f(y,x.l||hljs.IR,true);for(var w in x.k){if(!x.k.hasOwnProperty(w)){continue}if(x.k[w] instanceof Object){u=x.k[w]}else{u=x.k;w="keyword"}for(var r in u){if(!u.hasOwnProperty(r)){continue}x.k[r]=[w,u[r]];s.push(r)}}}if(!v){if(x.bWK){x.b="\\b("+s.join("|")+")\\s"}x.bR=f(y,x.b?x.b:"\\B|\\b");if(!x.e&&!x.eW){x.e="\\B|\\b"}if(x.e){x.eR=f(y,x.e)}}if(x.i){x.iR=f(y,x.i)}if(x.r===undefined){x.r=1}if(!x.c){x.c=[]}x.compiled=true;for(var t=0;t<x.c.length;t++){if(x.c[t]=="self"){x.c[t]=x}q(x.c[t],y,false)}if(x.starts){q(x.starts,y,false)}}for(var p in e){if(!e.hasOwnProperty(p)){continue}q(e[p].dM,e[p],true)}}function d(B,C){if(!j.called){j();j.called=true}function q(r,M){for(var L=0;L<M.c.length;L++){if((M.c[L].bR.exec(r)||[null])[0]==r){return M.c[L]}}}function v(L,r){if(D[L].e&&D[L].eR.test(r)){return 1}if(D[L].eW){var M=v(L-1,r);return M?M+1:0}return 0}function w(r,L){return L.i&&L.iR.test(r)}function K(N,O){var M=[];for(var L=0;L<N.c.length;L++){M.push(N.c[L].b)}var r=D.length-1;do{if(D[r].e){M.push(D[r].e)}r--}while(D[r+1].eW);if(N.i){M.push(N.i)}return f(O,M.join("|"),true)}function p(M,L){var N=D[D.length-1];if(!N.t){N.t=K(N,E)}N.t.lastIndex=L;var r=N.t.exec(M);return r?[M.substr(L,r.index-L),r[0],false]:[M.substr(L),"",true]}function z(N,r){var L=E.cI?r[0].toLowerCase():r[0];var M=N.k[L];if(M&&M instanceof Array){return M}return false}function F(L,P){L=m(L);if(!P.k){return L}var r="";var O=0;P.lR.lastIndex=0;var M=P.lR.exec(L);while(M){r+=L.substr(O,M.index-O);var N=z(P,M);if(N){x+=N[1];r+='<span class="'+N[0]+'">'+M[0]+"</span>"}else{r+=M[0]}O=P.lR.lastIndex;M=P.lR.exec(L)}return r+L.substr(O,L.length-O)}function J(L,M){if(M.sL&&e[M.sL]){var r=d(M.sL,L);x+=r.keyword_count;return r.value}else{return F(L,M)}}function I(M,r){var L=M.cN?'<span class="'+M.cN+'">':"";if(M.rB){y+=L;M.buffer=""}else{if(M.eB){y+=m(r)+L;M.buffer=""}else{y+=L;M.buffer=r}}D.push(M);A+=M.r}function G(N,M,Q){var R=D[D.length-1];if(Q){y+=J(R.buffer+N,R);return false}var P=q(M,R);if(P){y+=J(R.buffer+N,R);I(P,M);return P.rB}var L=v(D.length-1,M);if(L){var O=R.cN?"</span>":"";if(R.rE){y+=J(R.buffer+N,R)+O}else{if(R.eE){y+=J(R.buffer+N,R)+O+m(M)}else{y+=J(R.buffer+N+M,R)+O}}while(L>1){O=D[D.length-2].cN?"</span>":"";y+=O;L--;D.length--}var r=D[D.length-1];D.length--;D[D.length-1].buffer="";if(r.starts){I(r.starts,"")}return R.rE}if(w(M,R)){throw"Illegal"}}var E=e[B];var D=[E.dM];var A=0;var x=0;var y="";try{var s,u=0;E.dM.buffer="";do{s=p(C,u);var t=G(s[0],s[1],s[2]);u+=s[0].length;if(!t){u+=s[1].length}}while(!s[2]);if(D.length>1){throw"Illegal"}return{r:A,keyword_count:x,value:y}}catch(H){if(H=="Illegal"){return{r:0,keyword_count:0,value:m(C)}}else{throw H}}}function g(t){var p={keyword_count:0,r:0,value:m(t)};var r=p;for(var q in e){if(!e.hasOwnProperty(q)){continue}var s=d(q,t);s.language=q;if(s.keyword_count+s.r>r.keyword_count+r.r){r=s}if(s.keyword_count+s.r>p.keyword_count+p.r){r=p;p=s}}if(r.language){p.second_best=r}return p}function i(r,q,p){if(q){r=r.replace(/^((<[^>]+>|\t)+)/gm,function(t,w,v,u){return w.replace(/\t/g,q)})}if(p){r=r.replace(/\n/g,"<br>")}return r}function n(t,w,r){var x=h(t,r);var v=a(t);var y,s;if(v){y=d(v,x)}else{return}var q=c(t);if(q.length){s=document.createElement("pre");s.innerHTML=y.value;y.value=k(q,c(s),x)}y.value=i(y.value,w,r);var u=t.className;if(!u.match("(\\s|^)(language-)?"+v+"(\\s|$)")){u=u?(u+" "+v):v}if(/MSIE [678]/.test(navigator.userAgent)&&t.tagName=="CODE"&&t.parentNode.tagName=="PRE"){s=t.parentNode;var p=document.createElement("div");p.innerHTML="<pre><code>"+y.value+"</code></pre>";t=p.firstChild.firstChild;p.firstChild.cN=s.cN;s.parentNode.replaceChild(p.firstChild,s)}else{t.innerHTML=y.value}t.className=u;t.result={language:v,kw:y.keyword_count,re:y.r};if(y.second_best){t.second_best={language:y.second_best.language,kw:y.second_best.keyword_count,re:y.second_best.r}}}function o(){if(o.called){return}o.called=true;var r=document.getElementsByTagName("pre");for(var p=0;p<r.length;p++){var q=b(r[p]);if(q){n(q,hljs.tabReplace)}}}function l(){if(window.addEventListener){window.addEventListener("DOMContentLoaded",o,false);window.addEventListener("load",o,false)}else{if(window.attachEvent){window.attachEvent("onload",o)}else{window.onload=o}}}var e={};this.LANGUAGES=e;this.highlight=d;this.highlightAuto=g;this.fixMarkup=i;this.highlightBlock=n;this.initHighlighting=o;this.initHighlightingOnLoad=l;this.IR="[a-zA-Z][a-zA-Z0-9_]*";this.UIR="[a-zA-Z_][a-zA-Z0-9_]*";this.NR="\\b\\d+(\\.\\d+)?";this.CNR="\\b(0[xX][a-fA-F0-9]+|(\\d+(\\.\\d*)?|\\.\\d+)([eE][-+]?\\d+)?)";this.BNR="\\b(0b[01]+)";this.RSR="!|!=|!==|%|%=|&|&&|&=|\\*|\\*=|\\+|\\+=|,|\\.|-|-=|/|/=|:|;|<|<<|<<=|<=|=|==|===|>|>=|>>|>>=|>>>|>>>=|\\?|\\[|\\{|\\(|\\^|\\^=|\\||\\|=|\\|\\||~";this.ER="(?![\\s\\S])";this.BE={b:"\\\\.",r:0};this.ASM={cN:"string",b:"'",e:"'",i:"\\n",c:[this.BE],r:0};this.QSM={cN:"string",b:'"',e:'"',i:"\\n",c:[this.BE],r:0};this.CLCM={cN:"comment",b:"//",e:"$"};this.CBLCLM={cN:"comment",b:"/\\*",e:"\\*/"};this.HCM={cN:"comment",b:"#",e:"$"};this.NM={cN:"number",b:this.NR,r:0};this.CNM={cN:"number",b:this.CNR,r:0};this.BNM={cN:"number",b:this.BNR,r:0};this.inherit=function(r,s){var p={};for(var q in r){p[q]=r[q]}if(s){for(var q in s){p[q]=s[q]}}return p}}();hljs.LANGUAGES.cpp=function(){var a={keyword:{"false":1,"int":1,"float":1,"while":1,"private":1,"char":1,"catch":1,"export":1,virtual:1,operator:2,sizeof:2,dynamic_cast:2,typedef:2,const_cast:2,"const":1,struct:1,"for":1,static_cast:2,union:1,namespace:1,unsigned:1,"long":1,"throw":1,"volatile":2,"static":1,"protected":1,bool:1,template:1,mutable:1,"if":1,"public":1,friend:2,"do":1,"return":1,"goto":1,auto:1,"void":2,"enum":1,"else":1,"break":1,"new":1,extern:1,using:1,"true":1,"class":1,asm:1,"case":1,typeid:1,"short":1,reinterpret_cast:2,"default":1,"double":1,register:1,explicit:1,signed:1,typename:1,"try":1,"this":1,"switch":1,"continue":1,wchar_t:1,inline:1,"delete":1,alignof:1,char16_t:1,char32_t:1,constexpr:1,decltype:1,noexcept:1,nullptr:1,static_assert:1,thread_local:1,restrict:1,_Bool:1,complex:1},built_in:{std:1,string:1,cin:1,cout:1,cerr:1,clog:1,stringstream:1,istringstream:1,ostringstream:1,auto_ptr:1,deque:1,list:1,queue:1,stack:1,vector:1,map:1,set:1,bitset:1,multiset:1,multimap:1,unordered_set:1,unordered_map:1,unordered_multiset:1,unordered_multimap:1,array:1,shared_ptr:1}};return{dM:{k:a,i:"</",c:[hljs.CLCM,hljs.CBLCLM,hljs.QSM,{cN:"string",b:"'\\\\?.",e:"'",i:"."},{cN:"number",b:"\\b(\\d+(\\.\\d*)?|\\.\\d+)(u|U|l|L|ul|UL|f|F)"},hljs.CNM,{cN:"preprocessor",b:"#",e:"$"},{cN:"stl_container",b:"\\b(deque|list|queue|stack|vector|map|set|bitset|multiset|multimap|unordered_map|unordered_set|unordered_multiset|unordered_multimap|array)\\s*<",e:">",k:a,r:10,c:["self"]}]}}}();hljs.LANGUAGES.r={dM:{c:[hljs.HCM,{cN:"number",b:"\\b0[xX][0-9a-fA-F]+[Li]?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+(?:[eE][+\\-]?\\d*)?L\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+\\.(?!\\d)(?:i\\b)?",e:hljs.IMMEDIATE_RE,r:1},{cN:"number",b:"\\b\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"keyword",b:"(?:tryCatch|library|setGeneric|setGroupGeneric)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\.",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\d+(?![\\w.])",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\b(?:function)",e:hljs.IMMEDIATE_RE,r:2},{cN:"keyword",b:"(?:if|in|break|next|repeat|else|for|return|switch|while|try|stop|warning|require|attach|detach|source|setMethod|setClass)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"literal",b:"(?:NA|NA_integer_|NA_real_|NA_character_|NA_complex_)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"literal",b:"(?:NULL|TRUE|FALSE|T|F|Inf|NaN)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"identifier",b:"[a-zA-Z.][a-zA-Z0-9._]*\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"<\\-(?!\\s*\\d)",e:hljs.IMMEDIATE_RE,r:2},{cN:"operator",b:"\\->|<\\-",e:hljs.IMMEDIATE_RE,r:1},{cN:"operator",b:"%%|~",e:hljs.IMMEDIATE_RE},{cN:"operator",b:">=|<=|==|!=|\\|\\||&&|=|\\+|\\-|\\*|/|\\^|>|<|!|&|\\||\\$|:",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"%",e:"%",i:"\\n",r:1},{cN:"identifier",b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[hljs.BE],r:0},{cN:"string",b:"'",e:"'",c:[hljs.BE],r:0},{cN:"paren",b:"[[({\\])}]",e:hljs.IMMEDIATE_RE,r:0}]}};
hljs.initHighlightingOnLoad();
</script>



<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 13px;
}

body {
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 20px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 {
   font-size:2.2em;
}

h2 {
   font-size:1.8em;
}

h3 {
   font-size:1.4em;
}

h4 {
   font-size:1.0em;
}

h5 {
   font-size:0.9em;
}

h6 {
   font-size:0.8em;
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre, img {
  max-width: 100%;
}
pre {
  overflow-x: auto;
}
pre code {
   display: block; padding: 0.5em;
}

code {
  font-size: 92%;
  border: 1px solid #ccc;
}

code[class] {
  background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * {
      background: transparent !important;
      color: black !important;
      filter:none !important;
      -ms-filter: none !important;
   }

   body {
      font-size:12pt;
      max-width:100%;
   }

   a, a:visited {
      text-decoration: underline;
   }

   hr {
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote {
      padding-right: 1em;
      page-break-inside: avoid;
   }

   tr, img {
      page-break-inside: avoid;
   }

   img {
      max-width: 100% !important;
   }

   @page :left {
      margin: 15mm 20mm 15mm 10mm;
   }

   @page :right {
      margin: 15mm 10mm 15mm 20mm;
   }

   p, h2, h3 {
      orphans: 3; widows: 3;
   }

   h2, h3 {
      page-break-after: avoid;
   }
}
</style>



</head>

<body>
<!-- library(rmarkdown) ; render("format-data.Rmd",html_vignette(toc=TRUE)) -->

<h2>Format data</h2>

<p>This document describes the format of the data used in 
a <code>conStruct</code> analysis.</p>

<p>For information on how to run a <code>conStruct</code> analysis 
after you&#39;ve formatted your data, see the companion 
vignette on <a href="run-conStruct.html">how to run conStruct</a>.</p>

<p>Throughout the document, I&#39;ll be referring to the 
example dataset included with the package:</p>

<pre><code class="r">library(conStruct)
#&gt; Loading required package: Rcpp
data(conStruct.data)
</code></pre>

<h2>conStruct data</h2>

<p>There are 3 data objects you need to run a <code>conStruct</code> analysis:</p>

<ol>
<li><p>[allele frequency data]</p></li>
<li><p>[geographic sampling coordinates]</p></li>
<li><p>[geographic distance matrix]</p></li>
</ol>

<p>In the sections below, I walk through the specific format required for each.</p>

<h3>Allele frequency data</h3>

<p>You must specify a matrix of allele frequency data for your samples.
(Make sure the data are of class <code>matrix</code>, and that it&#39;s not a <code>data.frame</code>.)
I assume that the data consist of bi-allelic SNPs.
At each locus, you pick an allele to count across all samples
(it doesn&#39;t matter whether it&#39;s randomly chosen or 
whether it&#39;s always the major or minor allele).
The frequency of the counted allele at a locus in a sample 
is the number of times the counted allele is observed at a locus 
divided by the number of chromosomes genotyped in that sample.
A sample can consist of a single individual or multiple individuals 
lumped together.
So, a successfully genotyped diploid individual heterozygous at a
particular locus would have an allele frequency of 0.5.
If the sample is a population of 13 haploids, of which 12 have the 
counted allele at a given locus, the frequency in that sample at that 
locus would be 12/13.</p>

<p>The matrix of allele frequencies should have one row per sample and 
one column per locus.  Missing data should be denoted with the value <code>NA</code>.
An small example allele frequency data matrix is shown below:</p>

<table><thead>
<tr>
<th align="left">Sample</th>
<th align="center">Locus1</th>
<th align="center">Locus2</th>
<th align="center">Locus3</th>
<th align="center">Locus4</th>
<th align="center">Locus5</th>
<th align="center">Locus6</th>
<th align="center">Locus7</th>
<th align="center">Locus8</th>
<th align="center">Locus9</th>
<th align="center">Locus10</th>
</tr>
</thead><tbody>
<tr>
<td align="left">Pop1</td>
<td align="center">0</td>
<td align="center">1</td>
<td align="center">NA</td>
<td align="center">0.8</td>
<td align="center">0.7</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0.6</td>
<td align="center">0</td>
<td align="center">1</td>
</tr>
<tr>
<td align="left">Pop2</td>
<td align="center">0</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">0.9</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">0.1</td>
<td align="center">0.6</td>
<td align="center">0</td>
<td align="center">0.9</td>
</tr>
<tr>
<td align="left">Pop3</td>
<td align="center">0.2</td>
<td align="center">0.75</td>
<td align="center">0</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">NA</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">0.1</td>
<td align="center">1</td>
</tr>
<tr>
<td align="left">Pop4</td>
<td align="center">0.1</td>
<td align="center">0.9</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">0.8</td>
<td align="center">1</td>
<td align="center">0.2</td>
<td align="center">0.7</td>
<td align="center">0.1</td>
<td align="center">0.3</td>
</tr>
<tr>
<td align="left">Pop5</td>
<td align="center">0</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">0.3</td>
<td align="center">0.9</td>
<td align="center">0.3</td>
<td align="center">NA</td>
</tr>
</tbody></table>

<p>An full example allele frequency data matrix is included in the 
<code>conStruct.data</code> object included with the package.</p>

<pre><code class="r"># load the example data object
data(conStruct.data)

# look at the allele frequency data 
#   for the first 5 populations and 10 loci
conStruct.data$allele.frequencies[1:5,1:10]
#&gt;         loc1 loc2 loc3 loc4 loc5 loc6 loc7 loc8 loc9 loc10
#&gt; sample1 0.00    0    0 0.00    0    0  0.0 0.00 0.80   0.0
#&gt; sample2 0.00    0    0 0.00    0    0  0.0 0.00 0.20   0.0
#&gt; sample3 0.00    0    0 0.05    0    0  0.0 0.05 0.05   0.0
#&gt; sample4 0.00    0    0 0.05    0    0  0.0 0.00 0.05   0.0
#&gt; sample7 0.65    0    0 0.00    0    0  0.1 0.00 0.20   0.8
</code></pre>

<h3>Geographic sampling coordinates</h3>

<p>You must specify a matrix of geographic sampling coordinates, 
which will be used for plotting the results of the analysis.
This should be a matrix with two columns that give the sample 
x-coordinates (longitude) and y-coordinates (latitude), 
respectively.  The order of rows of the matrix should be the same 
as the order of the rows of the allele frequencies matrix.
If you specify longitude and latitude, they should be in 
decimal degrees.</p>

<p>A full example sampling coordinate data matrix is included 
in the <code>conStruct.data</code> object included with the package.</p>

<pre><code class="r"># load the example data object
data(conStruct.data)

# look at the geographic sampling coordinates 
#   for the first 5 populations
conStruct.data$coords[1:5,]
#&gt;           Lon     Lat
#&gt; [1,] -84.4328 42.7828
#&gt; [2,] -84.4168 42.7828
#&gt; [3,] -84.4008 42.7828
#&gt; [4,] -84.3848 42.7828
#&gt; [5,] -84.4328 42.7988
</code></pre>

<h3>Geographic distance matrix</h3>

<p>If you choose to run the spatial model implemented in <code>conStruct</code>, 
you must specify a matrix of pairwise geographic distance between 
all samples.  If the coordinates of the samples are real locations 
on Earth (as opposed to simulated coordinates), I recommend 
calculating pairwise great-circle distance between sampling 
coordinates (using, e.g., <code>fields::rdist.earth</code>).</p>

<p>The order of the samples in the geographic distance matrix should 
match both that of the geographic coordinates and that of the 
allele frequency data matrix, and all three matrices should have 
the same number of rows.</p>

<p>The geographic distance matrix you specify should be the full 
matrix (that is, not the upper- or lower-triangles), with values 
of <code>0</code> on the diagonal entries.</p>

<p>A full example geographic distance matrix between all samples 
is in the <code>conStruct.data</code> object included with the package.</p>

<pre><code class="r"># load the example data object
data(conStruct.data)

# look at pariwise geographic distance 
#   between the first 5 populations
conStruct.data$geoDist[1:5,1:5]
#&gt;           [,1]      [,2]      [,3]      [,4]     [,5]
#&gt; [1,] 0.0000000 0.8122983 1.6245967 2.4368950 1.106773
#&gt; [2,] 0.8122983 0.0000000 0.8122983 1.6245967 1.372809
#&gt; [3,] 1.6245967 0.8122983 0.0000000 0.8122983 1.965599
#&gt; [4,] 2.4368950 1.6245967 0.8122983 0.0000000 2.676167
#&gt; [5,] 1.1067733 1.3728092 1.9655990 2.6761670 0.000000
</code></pre>

<h1>Other formats to conStruct</h1>

<p>For convenience, I&#39;ve written a function to convert population 
genetic data in STRUCTURE format to that used in <code>conStruct</code>.</p>

<h2>STRUCTURE to conStruct</h2>

<p>The program STRUCTURE is one of the most widely used 
methods for model-based clustering in population genetics. 
Many existing programs, including plink (v1.9 and above) 
and PgdSpider, convert data from diverse formats (including 
.vcf files) into STRUCTURE format.  In this section of the 
vignette, I walk through an example of converting a STRUCTURE 
format data file into a <code>conStruct</code> format data file.</p>

<h3>STRUCTURE data format</h3>

<p>More extensive documentation on STRUCTURE&#39;s data format 
can be found in <a href="https://web.stanford.edu/group/pritchardlab/structure_software/release_versions/v2.3.4/structure_doc.pdf">the STRUCTURE manual</a>.
An example STRUCTURE-formatted dataset is shown below:</p>

<table><thead>
<tr>
<th align="center"></th>
<th align="center"></th>
<th align="center">Loc1</th>
<th align="center">Loc1</th>
<th align="center">Loc2</th>
<th align="center">Loc2</th>
<th align="center">Loc3</th>
<th align="center">Loc3</th>
<th align="center">Loc4</th>
<th align="center">Loc4</th>
</tr>
</thead><tbody>
<tr>
<td align="center">Ind1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">1</td>
<td align="center">2</td>
<td align="center">-9</td>
<td align="center">-9</td>
</tr>
<tr>
<td align="center">Ind2</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
</tr>
<tr>
<td align="center">Ind3</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
</tr>
<tr>
<td align="center">Ind4</td>
<td align="center">2</td>
<td align="center">-9</td>
<td align="center">-9</td>
<td align="center">1</td>
<td align="center">2</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
</tr>
<tr>
<td align="center">Ind5</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">1</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">2</td>
</tr>
</tbody></table>

<pre><code>Example STRUCTURE format dataset, with one row per individual 
and two columns per locus. The first column gives sample names, the 
second refers to the sample locations, and the last 8 columns give 
genotype data for four loci. The numbers in the genotype data refer to 
the allele present at that locus: A1 = `1`, A2 = `2`, missing = `-9`.
</code></pre>

<p>To convert a STRUCTURE format file to <code>conStruct</code> format, 
you can use the function <code>structure2conStruct</code>, included in the 
<code>conStruct</code> package.</p>

<p>Below, I give an example of the usage of this function, assuming 
that the file containing the STRUCTURE format data is called 
&ldquo;myStructureData.str&rdquo;, and that it&#39;s on the &ldquo;desktop&rdquo; directory 
on the computer.  I also assume that the data are formatted as in 
the table above, with the genotype data starting at the 3rd column 
of the data matrix, and missing data denoted with a value of -9.</p>

<p><strong>Note that the STRUCTURE-format data must be a text file
and there can be no lines of text before the data table begins.
If your file is in an Excel spreadsheet, it can be converted to 
a text file using Save As &gt; File Format = Tab delimited Text 
(.txt). If there are lines of text at the top of the document 
before the data matrix begins, they must be deleted. In addition, 
your data can only contain bi-allelic data. If you have loci with 
more than two alleles, they should be not be included in the dataset. 
For more information on multi-allelic datasets, see the section on 
<a href="#microsatellites">Microsatellites</a> below.</strong></p>

<pre><code class="r">conStruct.data &lt;- structure2conStruct(infile = &quot;~/Desktop/myStructureData.str&quot;,
                                      onerowperind = TRUE,
                                      start.loci = 3,
                                      missing.datum = -9,
                                      outfile = &quot;~/Desktop/myConStructData&quot;)

</code></pre>

<p>An alternate STRUCTURE data format has two rows and one column per 
diploid genotype:</p>

<table><thead>
<tr>
<th align="center"></th>
<th align="center"></th>
<th align="center">Loc1</th>
<th align="center">Loc2</th>
<th align="center">Loc3</th>
<th align="center">Loc4</th>
<th align="center">Loc5</th>
<th align="center">Loc6</th>
<th align="center">Loc7</th>
<th align="center">Loc8</th>
</tr>
</thead><tbody>
<tr>
<td align="center">Ind1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">1</td>
<td align="center">2</td>
<td align="center">0</td>
<td align="center">1</td>
</tr>
<tr>
<td align="center">Ind1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">0</td>
<td align="center">1</td>
</tr>
<tr>
<td align="center">Ind2</td>
<td align="center">1</td>
<td align="center">0</td>
<td align="center">1</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
</tr>
<tr>
<td align="center">Ind2</td>
<td align="center">1</td>
<td align="center">0</td>
<td align="center">2</td>
<td align="center">1</td>
<td align="center">2</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
</tr>
<tr>
<td align="center">Ind3</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">1</td>
<td align="center">2</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">2</td>
</tr>
<tr>
<td align="center">Ind3</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">1</td>
<td align="center">2</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">2</td>
</tr>
<tr>
<td align="center">Ind4</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">0</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">1</td>
<td align="center">0</td>
<td align="center">2</td>
</tr>
<tr>
<td align="center">Ind4</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">0</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">0</td>
<td align="center">2</td>
</tr>
</tbody></table>

<pre><code>Example STRUCTURE format dataset, with two rows per individual 
and one column per locus. The first column gives sample names, the 
second refers to the sample locations, and the last 8 columns give 
genotype data for 8 loci. The numbers in the genotype data refer to 
the allele present at that locus: A1 = `1`, A2 = `2`, missing = `0`.
</code></pre>

<p>Data in this format can be converted to <code>conStruct</code> format using the 
command below:</p>

<pre><code class="r">conStruct.data &lt;- structure2conStruct(infile = &quot;~/Desktop/myStructureData.str&quot;,
                                      onerowperind = FALSE,
                                      start.loci = 3,
                                      missing.datum = 0,
                                      outfile = &quot;~/Desktop/myConStructData&quot;)

</code></pre>

<p>Further documentation for this function is in its help page, 
which you can go to using the command <code>help(structure2conStruct)</code>.</p>

<p>If you wish to group multiple individuals together into a single 
sample for analysis you can collapse rows of the <code>conStruct</code> format 
data file.  For example, if you have 12 individuals from 4 
locations (3 individuals from each location), and you wish to 
analyze the data treating populations at a sampling location 
as the unit of analysis,  you can do something like the
following:</p>

<pre><code class="r">pop.data.matrix &lt;- matrix(NA,nrow=4,ncol=ncol(conStruct.data))
for(i in 1:nrow(pop.data.matrix)){
    pop.data.matrix[i,] &lt;- colMeans(
                                conStruct.data[
                                    which(pop.index==i),,
                                    drop=FALSE
                                ],na.rm=TRUE
                            )
}
</code></pre>

<p>where <code>pop.index</code> is a vector that gives the population of origin 
for each of the individuals sampled.  In the example above, with 
12 individuals sampled from 4 locations (3 from each), 
<code>pop.index</code> would be <code>c(1,1,1,2,2,2,3,3,3,4,4,4)</code>.</p>

<h2>Microsatellites</h2>

<p>This method is designed to run on large datasets consisting of 
bi-allelic SNPs.  If you have a microsatellite dataset and you 
wish to run <code>conStruct</code>, the first consideration is whether you 
have sufficient data.  You should have more loci than samples 
in your data matrix (i.e., your data matrix should have more 
columns than rows).</p>

<p>If that&#39;s the case, the second consideration is how to format your 
microsat data so that you can run conStruct.  There are two standard 
ways of &ldquo;SNP-ifying&rdquo; a microsat dataset.</p>

<p>The first is to lump all microsatellite alleles present at a locus 
into two categories: &ldquo;major&rdquo; and &ldquo;other&rdquo;.  The &ldquo;major&rdquo; allele is the 
allele that occurs most frequently at a particular locus; all other 
alleles are put in the &ldquo;other&rdquo; bin.  You then can create a dataset 
in which you only report the frequency of the major allele, 
effectively reducing the number of alleles per locus to 2. 
This method has the disadvantage of throwing out data, but 
acknowledges the simplex relationships between alleles at a locus 
(the sum of the frequencies of all alleles at a locus must be 1).</p>

<p>The second approach, introduced by Cavalli-Sforza, is to split out 
each allele at a locus into a separate pseudo-locus consisting of only 
that allele.  That is, if you had 4 alleles present in the genotyped 
sample at a particular locus, at frequencies {0.4,0.3,0.1,0.2}, 
you would split those out into 4 separate columns in your data matrix 
(pseudo-loci), with frequencies in the sampled population of 
{0.4,0.3,0.1,0.2}.  This approach has the advantage of not throwing 
data away, but does not acknowledge the inter-allele dependence 
structure in frequencies, and therefore introduces some 
pseudoreplication into the dataset. This pseudoreplication may make 
you overconfident in your results, as the credible intervals on 
parameter estimates may be artificially narrow.</p>

<p>I would recommend trying both approaches, and comparing the 
estimates of pairwise relatedness you get from each to those 
derived from the raw microsatellite data to see which best 
recovers the patterns of relatedness in the data.  I also recommend 
running <code>conStruct</code> on datasets SNP-ified using both approaches, 
and comparing the results.</p>

</body>

</html>
