<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Gideon Bradburd" />

<meta name="date" content="2024-01-08" />

<title>How to format data for a conStruct analysis</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">How to format data for a conStruct
analysis</h1>
<h4 class="author">Gideon Bradburd</h4>
<h4 class="date">January 08, 2024</h4>


<div id="TOC">
<ul>
<li><a href="#format-data" id="toc-format-data">Format data</a></li>
<li><a href="#construct-data" id="toc-construct-data">conStruct data</a>
<ul>
<li><a href="#allele-frequency-data" id="toc-allele-frequency-data">Allele frequency data</a></li>
<li><a href="#geographic-sampling-coordinates" id="toc-geographic-sampling-coordinates">Geographic sampling
coordinates</a></li>
<li><a href="#geographic-distance-matrix" id="toc-geographic-distance-matrix">Geographic distance matrix</a></li>
</ul></li>
<li><a href="#other-formats-to-construct" id="toc-other-formats-to-construct">Other formats to conStruct</a>
<ul>
<li><a href="#structure-to-construct" id="toc-structure-to-construct">STRUCTURE to conStruct</a>
<ul>
<li><a href="#structure-data-format" id="toc-structure-data-format">STRUCTURE data format</a></li>
</ul></li>
<li><a href="#microsatellites" id="toc-microsatellites">Microsatellites</a></li>
</ul></li>
</ul>
</div>

<!-- library(rmarkdown) ; render("format-data.Rmd",html_vignette(toc=TRUE)) -->
<div id="format-data" class="section level2">
<h2>Format data</h2>
<p>This document describes the format of the data used in a
<code>conStruct</code> analysis.</p>
<p>For information on how to run a <code>conStruct</code> analysis after
you’ve formatted your data, see the companion vignette on <a href="run-conStruct.html">how to run conStruct</a>.</p>
<p>Throughout the document, I’ll be referring to the example dataset
included with the package:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(conStruct)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="co">#&gt; Loading required package: Rcpp</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">data</span>(conStruct.data)</span></code></pre></div>
</div>
<div id="construct-data" class="section level2">
<h2>conStruct data</h2>
<p>There are 3 data objects you need to run a <code>conStruct</code>
analysis:</p>
<ol style="list-style-type: decimal">
<li><p><a href="#allele-frequency-data">allele frequency
data</a></p></li>
<li><p><a href="#geographic-sampling-coordinates">geographic sampling
coordinates</a></p></li>
<li><p><a href="#geographic-distance-matrix">geographic distance
matrix</a></p></li>
</ol>
<p>In the sections below, I walk through the specific format required
for each.</p>
<div id="allele-frequency-data" class="section level3">
<h3>Allele frequency data</h3>
<p>You must specify a matrix of allele frequency data for your samples.
(Make sure the data are of class <code>matrix</code>, and that it’s not
a <code>data.frame</code>.) I assume that the data consist of bi-allelic
SNPs. At each locus, you pick an allele to count across all samples (it
doesn’t matter whether it’s randomly chosen or whether it’s always the
major or minor allele). The frequency of the counted allele at a locus
in a sample is the number of times the counted allele is observed at a
locus divided by the number of chromosomes genotyped in that sample. A
sample can consist of a single individual or multiple individuals lumped
together. So, a successfully genotyped diploid individual heterozygous
at a particular locus would have an allele frequency of 0.5. If the
sample is a population of 13 haploids, of which 12 have the counted
allele at a given locus, the frequency in that sample at that locus
would be 12/13.</p>
<p>The matrix of allele frequencies should have one row per sample and
one column per locus. Missing data should be denoted with the value
<code>NA</code>. An small example allele frequency data matrix is shown
below:</p>
<table>
<colgroup>
<col width="10%" />
<col width="8%" />
<col width="8%" />
<col width="8%" />
<col width="8%" />
<col width="8%" />
<col width="8%" />
<col width="8%" />
<col width="8%" />
<col width="8%" />
<col width="8%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">Sample</th>
<th align="center">Locus1</th>
<th align="center">Locus2</th>
<th align="center">Locus3</th>
<th align="center">Locus4</th>
<th align="center">Locus5</th>
<th align="center">Locus6</th>
<th align="center">Locus7</th>
<th align="center">Locus8</th>
<th align="center">Locus9</th>
<th align="center">Locus10</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Pop1</td>
<td align="center">0</td>
<td align="center">1</td>
<td align="center">NA</td>
<td align="center">0.8</td>
<td align="center">0.7</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0.6</td>
<td align="center">0</td>
<td align="center">1</td>
</tr>
<tr class="even">
<td align="left">Pop2</td>
<td align="center">0</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">0.9</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">0.1</td>
<td align="center">0.6</td>
<td align="center">0</td>
<td align="center">0.9</td>
</tr>
<tr class="odd">
<td align="left">Pop3</td>
<td align="center">0.2</td>
<td align="center">0.75</td>
<td align="center">0</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">NA</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">0.1</td>
<td align="center">1</td>
</tr>
<tr class="even">
<td align="left">Pop4</td>
<td align="center">0.1</td>
<td align="center">0.9</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">0.8</td>
<td align="center">1</td>
<td align="center">0.2</td>
<td align="center">0.7</td>
<td align="center">0.1</td>
<td align="center">0.3</td>
</tr>
<tr class="odd">
<td align="left">Pop5</td>
<td align="center">0</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">0.3</td>
<td align="center">0.9</td>
<td align="center">0.3</td>
<td align="center">NA</td>
</tr>
</tbody>
</table>
<p>An full example allele frequency data matrix is included in the
<code>conStruct.data</code> object included with the package.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># load the example data object</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="fu">data</span>(conStruct.data)</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="co"># look at the allele frequency data </span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="co">#   for the first 5 populations and 10 loci</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>conStruct.data<span class="sc">$</span>allele.frequencies[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="co">#&gt;         loc1 loc2 loc3 loc4 loc5 loc6 loc7 loc8 loc9 loc10</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="co">#&gt; sample1 0.00    0    0 0.00    0    0  0.0 0.00 0.80   0.0</span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a><span class="co">#&gt; sample2 0.00    0    0 0.00    0    0  0.0 0.00 0.20   0.0</span></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="co">#&gt; sample3 0.00    0    0 0.05    0    0  0.0 0.05 0.05   0.0</span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a><span class="co">#&gt; sample4 0.00    0    0 0.05    0    0  0.0 0.00 0.05   0.0</span></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a><span class="co">#&gt; sample7 0.65    0    0 0.00    0    0  0.1 0.00 0.20   0.8</span></span></code></pre></div>
</div>
<div id="geographic-sampling-coordinates" class="section level3">
<h3>Geographic sampling coordinates</h3>
<p>You must specify a matrix of geographic sampling coordinates, which
will be used for plotting the results of the analysis. This should be a
matrix with two columns that give the sample x-coordinates (longitude)
and y-coordinates (latitude), respectively. The order of rows of the
matrix should be the same as the order of the rows of the allele
frequencies matrix. If you specify longitude and latitude, they should
be in decimal degrees.</p>
<p>A full example sampling coordinate data matrix is included in the
<code>conStruct.data</code> object included with the package.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># load the example data object</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="fu">data</span>(conStruct.data)</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="co"># look at the geographic sampling coordinates </span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="co">#   for the first 5 populations</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>conStruct.data<span class="sc">$</span>coords[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,]</span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="co">#&gt;           Lon     Lat</span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a><span class="co">#&gt; [1,] -84.4328 42.7828</span></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a><span class="co">#&gt; [2,] -84.4168 42.7828</span></span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a><span class="co">#&gt; [3,] -84.4008 42.7828</span></span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a><span class="co">#&gt; [4,] -84.3848 42.7828</span></span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a><span class="co">#&gt; [5,] -84.4328 42.7988</span></span></code></pre></div>
</div>
<div id="geographic-distance-matrix" class="section level3">
<h3>Geographic distance matrix</h3>
<p>If you choose to run the spatial model implemented in
<code>conStruct</code>, you must specify a matrix of pairwise geographic
distance between all samples. If the coordinates of the samples are real
locations on Earth (as opposed to simulated coordinates), I recommend
calculating pairwise great-circle distance between sampling coordinates
(using, e.g., <code>geosphere::distm</code> or
<code>geosphere::distGeo</code>).</p>
<p>The order of the samples in the geographic distance matrix should
match both that of the geographic coordinates and that of the allele
frequency data matrix, and all three matrices should have the same
number of rows.</p>
<p>The geographic distance matrix you specify should be the full matrix
(that is, not the upper- or lower-triangles), with values of
<code>0</code> on the diagonal entries.</p>
<p>A full example geographic distance matrix between all samples is in
the <code>conStruct.data</code> object included with the package.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co"># load the example data object</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="fu">data</span>(conStruct.data)</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="co"># look at pariwise geographic distance </span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="co">#   between the first 5 populations</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>conStruct.data<span class="sc">$</span>geoDist[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a><span class="co">#&gt;           [,1]      [,2]      [,3]      [,4]     [,5]</span></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a><span class="co">#&gt; [1,] 0.0000000 0.8122983 1.6245967 2.4368950 1.106773</span></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a><span class="co">#&gt; [2,] 0.8122983 0.0000000 0.8122983 1.6245967 1.372809</span></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a><span class="co">#&gt; [3,] 1.6245967 0.8122983 0.0000000 0.8122983 1.965599</span></span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a><span class="co">#&gt; [4,] 2.4368950 1.6245967 0.8122983 0.0000000 2.676167</span></span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a><span class="co">#&gt; [5,] 1.1067733 1.3728092 1.9655990 2.6761670 0.000000</span></span></code></pre></div>
</div>
</div>
<div id="other-formats-to-construct" class="section level1">
<h1>Other formats to conStruct</h1>
<p>For convenience, I’ve written a function to convert population
genetic data in STRUCTURE format to that used in
<code>conStruct</code>.</p>
<div id="structure-to-construct" class="section level2">
<h2>STRUCTURE to conStruct</h2>
<p>The program STRUCTURE is one of the most widely used methods for
model-based clustering in population genetics. Many existing programs,
including plink (v1.9 and above) and PgdSpider, convert data from
diverse formats (including .vcf files) into STRUCTURE format. In this
section of the vignette, I walk through an example of converting a
STRUCTURE format data file into a <code>conStruct</code> format data
file.</p>
<div id="structure-data-format" class="section level3">
<h3>STRUCTURE data format</h3>
<p>More extensive documentation on STRUCTURE’s data format can be found
in <a href="https://web.stanford.edu/group/pritchardlab/structure_software/release_versions/v2.3.4/structure_doc.pdf">the
STRUCTURE manual</a>. An example STRUCTURE-formatted dataset is shown
below:</p>
<table>
<thead>
<tr class="header">
<th align="center"></th>
<th align="center"></th>
<th align="center">Loc1</th>
<th align="center">Loc1</th>
<th align="center">Loc2</th>
<th align="center">Loc2</th>
<th align="center">Loc3</th>
<th align="center">Loc3</th>
<th align="center">Loc4</th>
<th align="center">Loc4</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">Ind1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">1</td>
<td align="center">2</td>
<td align="center">-9</td>
<td align="center">-9</td>
</tr>
<tr class="even">
<td align="center">Ind2</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
</tr>
<tr class="odd">
<td align="center">Ind3</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
</tr>
<tr class="even">
<td align="center">Ind4</td>
<td align="center">2</td>
<td align="center">-9</td>
<td align="center">-9</td>
<td align="center">1</td>
<td align="center">2</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
</tr>
<tr class="odd">
<td align="center">Ind5</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">1</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">2</td>
</tr>
</tbody>
</table>
<pre><code>Example STRUCTURE format dataset, with one row per individual 
and two columns per locus. The first column gives sample names, the 
second refers to the sample locations, and the last 8 columns give 
genotype data for four loci. The numbers in the genotype data refer to 
the allele present at that locus: A1 = `1`, A2 = `2`, missing = `-9`.</code></pre>
<p>To convert a STRUCTURE format file to <code>conStruct</code> format,
you can use the function <code>structure2conStruct</code>, included in
the <code>conStruct</code> package.</p>
<p>Below, I give an example of the usage of this function, assuming that
the file containing the STRUCTURE format data is called
“myStructureData.str”, and that it’s on the “desktop” directory on the
computer. I also assume that the data are formatted as in the table
above, with the genotype data starting at the 3rd column of the data
matrix, and missing data denoted with a value of -9.</p>
<p><strong>Note that the STRUCTURE-format data must be a text file and
there can be no lines of text before the data table begins. If your file
is in an Excel spreadsheet, it can be converted to a text file using
Save As &gt; File Format = Tab delimited Text (.txt). If there are lines
of text at the top of the document before the data matrix begins, they
must be deleted or specified via the <code>start.samples</code>
argument. In addition, your data can only contain bi-allelic data. If
you have loci with more than two alleles, they should be not be included
in the dataset. For more information on multi-allelic datasets, see the
section on <a href="#microsatellites">Microsatellites</a>
below.</strong></p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>conStruct.data <span class="ot">&lt;-</span> <span class="fu">structure2conStruct</span>(<span class="at">infile =</span> <span class="st">&quot;~/Desktop/myStructureData.str&quot;</span>,</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>                                      <span class="at">onerowperind =</span> <span class="cn">TRUE</span>,</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>                                      <span class="at">start.loci =</span> <span class="dv">3</span>,</span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>                                      <span class="at">start.samples =</span> <span class="dv">1</span>,</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>                                      <span class="at">missing.datum =</span> <span class="sc">-</span><span class="dv">9</span>,</span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>                                      <span class="at">outfile =</span> <span class="st">&quot;~/Desktop/myConStructData&quot;</span>)</span></code></pre></div>
<p>An alternate STRUCTURE data format has two rows and one column per
diploid genotype:</p>
<table>
<thead>
<tr class="header">
<th align="center"></th>
<th align="center"></th>
<th align="center">Loc1</th>
<th align="center">Loc2</th>
<th align="center">Loc3</th>
<th align="center">Loc4</th>
<th align="center">Loc5</th>
<th align="center">Loc6</th>
<th align="center">Loc7</th>
<th align="center">Loc8</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">Ind1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">1</td>
<td align="center">2</td>
<td align="center">0</td>
<td align="center">1</td>
</tr>
<tr class="even">
<td align="center">Ind1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">0</td>
<td align="center">1</td>
</tr>
<tr class="odd">
<td align="center">Ind2</td>
<td align="center">1</td>
<td align="center">0</td>
<td align="center">1</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
</tr>
<tr class="even">
<td align="center">Ind2</td>
<td align="center">1</td>
<td align="center">0</td>
<td align="center">2</td>
<td align="center">1</td>
<td align="center">2</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
</tr>
<tr class="odd">
<td align="center">Ind3</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">1</td>
<td align="center">2</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">2</td>
</tr>
<tr class="even">
<td align="center">Ind3</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">1</td>
<td align="center">2</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">2</td>
</tr>
<tr class="odd">
<td align="center">Ind4</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">0</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">1</td>
<td align="center">0</td>
<td align="center">2</td>
</tr>
<tr class="even">
<td align="center">Ind4</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">0</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">0</td>
<td align="center">2</td>
</tr>
</tbody>
</table>
<pre><code>Example STRUCTURE format dataset, with two rows per individual 
and one column per locus. The first column gives sample names, the 
second refers to the sample locations, and the last 8 columns give 
genotype data for 8 loci. The numbers in the genotype data refer to 
the allele present at that locus: A1 = `1`, A2 = `2`, missing = `0`.</code></pre>
<p>Data in this format can be converted to <code>conStruct</code> format
using the command below:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>conStruct.data <span class="ot">&lt;-</span> <span class="fu">structure2conStruct</span>(<span class="at">infile =</span> <span class="st">&quot;~/Desktop/myStructureData.str&quot;</span>,</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>                                      <span class="at">onerowperind =</span> <span class="cn">FALSE</span>,</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>                                      <span class="at">start.loci =</span> <span class="dv">3</span>,</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>                                      <span class="at">start.samples =</span> <span class="dv">1</span>,</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a>                                      <span class="at">missing.datum =</span> <span class="dv">0</span>,</span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a>                                      <span class="at">outfile =</span> <span class="st">&quot;~/Desktop/myConStructData&quot;</span>)</span></code></pre></div>
<p>Further documentation for this function is in its help page, which
you can go to using the command
<code>help(structure2conStruct)</code>.</p>
<p>If you wish to group multiple individuals together into a single
sample for analysis you can collapse rows of the <code>conStruct</code>
format data file. For example, if you have 12 individuals from 4
locations (3 individuals from each location), and you wish to analyze
the data treating populations at a sampling location as the unit of
analysis, you can do something like the following:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>pop.data.matrix <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="cn">NA</span>,<span class="at">nrow=</span><span class="dv">4</span>,<span class="at">ncol=</span><span class="fu">ncol</span>(conStruct.data))</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(pop.data.matrix)){</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>    pop.data.matrix[i,] <span class="ot">&lt;-</span> <span class="fu">colMeans</span>(</span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>                                conStruct.data[</span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>                                    <span class="fu">which</span>(pop.index<span class="sc">==</span>i),,</span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a>                                    <span class="at">drop=</span><span class="cn">FALSE</span></span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a>                                ],<span class="at">na.rm=</span><span class="cn">TRUE</span></span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a>                            )</span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a>}</span></code></pre></div>
<p>where <code>pop.index</code> is a vector that gives the population of
origin for each of the individuals sampled. In the example above, with
12 individuals sampled from 4 locations (3 from each),
<code>pop.index</code> would be
<code>c(1,1,1,2,2,2,3,3,3,4,4,4)</code>.</p>
</div>
</div>
<div id="microsatellites" class="section level2">
<h2>Microsatellites</h2>
<p>This method is designed to run on large datasets consisting of
bi-allelic SNPs. If you have a microsatellite dataset and you wish to
run <code>conStruct</code>, the first consideration is whether you have
sufficient data. You should have more loci than samples in your data
matrix (i.e., your data matrix should have more columns than rows).</p>
<p>If that’s the case, the second consideration is how to format your
microsat data so that you can run conStruct. There are two standard ways
of “SNP-ifying” a microsat dataset.</p>
<p>The first is to lump all microsatellite alleles present at a locus
into two categories: “major” and “other”. The “major” allele is the
allele that occurs most frequently at a particular locus; all other
alleles are put in the “other” bin. You then can create a dataset in
which you only report the frequency of the major allele, effectively
reducing the number of alleles per locus to 2. This method has the
disadvantage of throwing out data, but acknowledges the simplex
relationships between alleles at a locus (the sum of the frequencies of
all alleles at a locus must be 1).</p>
<p>The second approach, introduced by Cavalli-Sforza, is to split out
each allele at a locus into a separate pseudo-locus consisting of only
that allele. That is, if you had 4 alleles present in the genotyped
sample at a particular locus, at frequencies {0.4,0.3,0.1,0.2}, you
would split those out into 4 separate columns in your data matrix
(pseudo-loci), with frequencies in the sampled population of
{0.4,0.3,0.1,0.2}. This approach has the advantage of not throwing data
away, but does not acknowledge the inter-allele dependence structure in
frequencies, and therefore introduces some pseudoreplication into the
dataset. This pseudoreplication may make you overconfident in your
results, as the credible intervals on parameter estimates may be
artificially narrow.</p>
<p>I would recommend trying both approaches, and comparing the estimates
of pairwise relatedness you get from each to those derived from the raw
microsatellite data to see which best recovers the patterns of
relatedness in the data. I also recommend running <code>conStruct</code>
on datasets SNP-ified using both approaches, and comparing the
results.</p>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
